Source/WebKit/UIProcess/API/gtk/WebKitPrintOperation.cpp

```diff
@@ -393,6 +424,222 @@ static void webkitPrintOperationPrintPagesForFrame(WebKitPrintOperation* printOp
+
+    PrintInfo printInfo(priv->printJob.get(), printOperation->priv->printMode);
+    auto& page = webkitWebViewGetPage(printOperation->priv->webView.get());
+    page.drawPagesForPrinting(webFrame, printInfo, [printOperation = GRefPtr<WebKitPrintOperation>(printOperation)](std::optional<WebCore::SharedMemory::Handle>&& data, WebCore::ResourceError&& error) mutable {
+        auto* priv = printOperation->priv;
+        // When running synchronously, WebPageProxy::printFrame() calls endPrinting().
+        if (priv->printMode == PrintInfo::PrintMode::Async && priv->webView)
+            webkitWebViewGetPage(priv->webView.get()).endPrinting();
+
+        if (!data || !error.isNull()) {
+            if (!error.isNull())
+                webkitPrintOperationFailed(printOperation.get(), WTFMove(error));
+            webkitPrintOperationFinished(printOperation.get());
+            return;
+        }
+
+        priv->data = data->releaseHandle();
+        GUniqueOutPtr<GError> jobError;
+        gtk_print_job_set_source_fd(priv->printJob.get(), priv->data.value(), &jobError.outPtr());
+        if (jobError) {
+            webkitPrintOperationFailed(printOperation.get(), GUniquePtr<GError> { g_error_new_literal(WEBKIT_PRINT_ERROR, WEBKIT_PRINT_ERROR_GENERAL, jobError->message) });
+            webkitPrintOperationFinished(printOperation.get());
+            return;
+        }
+
+        // "Print" the memory buffer into a temporary file, since the Print portal may take an
+        // arbitrary amount of time to finish the operation.
+        gtk_print_job_send(priv->printJob.get(), [](GtkPrintJob* printJob, gpointer userData, const GError* error) {
+            auto printOperation = adoptGRef(static_cast<WebKitPrintOperation*>(userData));
+            auto* priv = printOperation->priv;
+
+            // FIXME: this should be done in the parent callback
+            auto preparePrintResponse = std::exchange(priv->preparePrintResponse, std::nullopt);
+            RELEASE_ASSERT(preparePrintResponse && preparePrintResponse->token());
+
+            if (error) {
+                webkitPrintOperationFailed(printOperation.get(), GUniquePtr<GError> { g_error_new_literal(WEBKIT_PRINT_ERROR, WEBKIT_PRINT_ERROR_GENERAL, error->message) });
+                webkitPrintOperationFinished(printOperation.get());
+                return;
+            }
+
+            auto* printSettings = gtk_print_job_get_settings(printJob);
+            GUniquePtr<char> filename(g_filename_from_uri(gtk_print_settings_get(printSettings, GTK_PRINT_SETTINGS_OUTPUT_URI), NULL, NULL));
+            UnixFileDescriptor fd = UnixFileDescriptor { open(filename.get(), O_RDWR | O_CLOEXEC), UnixFileDescriptor::Adopt };
+            if (!fd) {
+                webkitPrintOperationFailed(printOperation.get(), GUniquePtr<GError> { g_error_new_literal(WEBKIT_PRINT_ERROR, WEBKIT_PRINT_ERROR_GENERAL, "Error opening temporary file") });
```

Comment:
```suggestion
                webkitPrintOperationFailed(printOperation.get(), GUniquePtr<GError> { g_error_new_literal(WEBKIT_PRINT_ERROR, WEBKIT_PRINT_ERROR_GENERAL, "Error opening temporary file") });
```

I would translate this error message too, since it's likely to show up in UIs. And we can make the error message more useful. Something closer to:

```suggestion
                webkitPrintOperationFailed(printOperation.get(), GUniquePtr<GError> { g_error_new(WEBKIT_PRINT_ERROR, WEBKIT_PRINT_ERROR_GENERAL, _("Error opening %s: %s"), filename.get(), safeStrerror(errno) });
```

We both spend too much time investigating odd bug reports as it is; might as well make it easier to figure out what is wrong when there is an error.